Lib/ast.py

```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
         with self.block(extra=self.get_type_comment(node)):
             self.traverse(node.body)
 
+    def _str_literal_helper(
+        self, string, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimizing escapes.
+        Returns (possible quote types, string literal to write).
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
+        escaped_string = "".join(
+            (
+                c.encode('unicode_escape').decode('ascii')
+                if c in escape or (not c.isprintable() and c not in '\n\t')
+                else c
+            )
+            for c in string
+        )
+        possible_quotes = [q for q in quote_types if q not in escaped_string]
+        if "\n" in escaped_string:
+            possible_quotes = [q for q in possible_quotes if q in ('"""', "'''")]
```

Comment:
Is it possible that we could somehow merge these 2 different iterations? (something like this?)
```diff
+        is_multi_line = "\n" in escaped_string
+
+        possible_quote_types = quote_types
+        if is_multi_line:
+            possible_quote_types = _MULTI_QUOTES
+        possible_quotes = [quote for quote in possible_quote_types if quote not in escaped_string]
```

Reply:
The suggestion isn't quite correct, since we'd need to make sure `possible_quote_types` is a subset of `quote_types`. We could do something like:
```
is_multi_line = "\n" in escaped_string
possible_quotes = [q for q in quote_types if (q not in escaped_string) and (not is_multi_line or q in _MULTI_QUOTES)]
```
But I think that's less readable. `len(quote_types) <= 4`, so removing an iteration isn't important.

Reply:
> we'd need to make sure possible_quote_types is a subset of quote_types

Makes sense, what about?
```diff
+        is_multi_line = "\n" in escaped_string
+
+        possible_quote_types = quote_types
+        if is_multi_line:
+            possible_quote_types = [quote for quote in possible_quote_types if quote in _MULTI_QUOTES]
+        possible_quotes = [quote for quote in possible_quote_types if quote not in escaped_string]
```

Reply:
Sure, I'll swap the order of the iterations. That could save us two iterations over the escaped string.